modules/layers/src/icon-layer/icon-manager.js

```diff
@@ -0,0 +1,259 @@
+/* eslint-disable */
+import GL from 'luma.gl/constants';
+import {Texture2D, loadImages, loadTextures} from 'luma.gl';
+
+const MAX_CANVAS_WIDTH = 1024;
+const MAX_CANVAS_HEIGHT = 768;
+
+const DEFAULT_TEXTURE_MIN_FILTER = GL.LINEAR_MIPMAP_LINEAR;
+// GL.LINEAR is the default value but explicitly set it here
+const DEFAULT_TEXTURE_MAG_FILTER = GL.LINEAR;
+
+const noop = () => {};
+
+function nextPowOfTwo(number) {
+  return Math.pow(2, Math.ceil(Math.log2(number)));
+}
+
+// traverse icons in a row of icon atlas
+// extend each icon with left-top coordinates
+function buildRowMapping(mapping, columns, yOffset) {
+  for (let i = 0; i < columns.length; i++) {
+    const {icon, xOffset} = columns[i];
+    mapping[icon.url] = Object.assign({}, icon, {
+      x: xOffset,
+      y: yOffset
+    });
+  }
+}
+
+function getIcons(data, getIcon) {
+  if (!data) {
+    return null;
+  }
+
+  return data.reduce((resMap, point) => {
+    const icon = getIcon(point);
+    if (!resMap[icon.url]) {
+      resMap[icon.url] = icon;
+    }
+    return resMap;
+  }, {});
+}
+
+// check if there are some icons from new data not fetched
+export function needRepack(oldData, data, getIcon) {
+  const oldIcons = getIcons(oldData, getIcon) || {};
```

Comment:
This can be very expensive. Imagine an app that animates `sizeScale` from 0 to 1. This will be called on every frame on potentially hundreds of thousands of objects.